File: 02/junyeong/longest-ideal-subsequence.rs
use std::collections::HashMap;

struct Solution {}

impl Solution {
    pub fn longest_ideal_string(s: String, k: i32) -> i32 {
        let mut map: HashMap<u8, i32> = HashMap::new();
        for byte in s.bytes() {
            let mut len = 1;
            for satisfied_byte in byte - (k as u8)..=byte + (k as u8) {
                len = len.max(*map.get(&satisfied_byte).unwrap_or(&0) + 1);
            }
            map.insert(byte, len);
        }
        *map.values().max().unwrap_or(&1)
    }
}

File: 02/byungkyu/longest-ideal-subsequence.cpp
class Solution {
public:
    int longestIdealString(string s, int k) {
        int dp[30];
        for(int j=0;j<30;j++){
            dp[j]=0;
        }
        for(int i=0;i<s.length();i++){
            int ch=s[i]-'a',ma=0;
            for(int j=max(0,ch-k);j<=min(25,ch+k);j++){
                ma=max(ma,dp[j]);
            }
            dp[ch]=max(dp[ch],ma+1);
		    }
        int ret=0;
        for(int i=0;i<26;i++)ret=max(ret,dp[i]);
        return ret;
        
    }
};